test/functional/example_test.py

```diff
@@ -0,0 +1,205 @@
+        NetworkThread().start()
+        # wait_for_verack ensures that the p2p connection is fully up.
+        node0.wait_for_verack()
+
+        # Generating a block on one of the nodes will get us out of IBD
+        blocks = [int(self.nodes[0].generate(nblocks=1)[0], 16)]
+
+        # Notice above how we called an RPC by calling a method with the same
+        # name on the node object. Notice also how we used a keyword argument
+        # to specify a named RPC argument. Neither of those are defined on the
+        # node object. Instead there's some __getattr__() magic going on under
+        # the covers to dispatch unrecognised attribute calls to the RPC
+        # interface.
+
+        # Logs are nice. Do plenty of them. They can be used in place of comments for
+        # breaking the test into sub-sections.
+        self.log.info("Starting test!")
+
+        self.log.info("Calling a custom function")
+        custom_function()
+
+        self.log.info("Calling a custom method")
+        self.custom_method()
+
+        self.log.info("Create some blocks")
+        self.tip = int(self.nodes[0].getbestblockhash(), 16)
+        self.block_time = self.nodes[0].getblock(self.nodes[0].getbestblockhash())['time'] + 1
+
+        height = 1
+
+        for i in range(10):
+            # Use the mininode and blocktools functionality to manually build a block
+            # Calling the generate() rpc is easier, but this allows us to exactly
+            # control the blocks and transactions.
+            block = create_block(self.tip, create_coinbase(height), self.block_time)
+            block.solve()
+            block_message = msg_block(block)
+            # Send message is used to send a p2p message to the node over our NodeConn connection
+            node0.send_message(block_message)
+            self.tip = block.sha256
+            blocks.append(self.tip)
+            self.block_time += 1
+            height += 1
+
+        self.log.info("Wait for node1 to reach current tip (height 11) using RPC")
+        self.nodes[1].waitforblockheight(11)
+
+        self.log.info("Connect node2 and node1")
+        connect_nodes_bi(self.nodes, 1, 2)
+
+        self.log.info("Add p2p connection to node2")
+        node2 = BaseNode()
+        connections.append(NodeConn('127.0.0.1', p2p_port(2), self.nodes[2], node2))
+        node2.add_connection(connections[1])
+        node2.wait_for_verack()
+
+        self.log.info("Wait for node2 reach current tip. Test that it has propogated all the blocks to us")
+
+        for block in blocks:
+            getdata_request = msg_getdata()
+            getdata_request.inv.append(CInv(2, block))
+            node2.send_message(getdata_request)
+
+        # wait_until() will loop until a predicate condition is met. Use it to test properties of the
+        # ConnNodeCB objects.
+        assert wait_until(lambda: sorted(blocks) == sorted(list(node2.block_receive_map.keys())), timeout=5)
```

Comment:
I think it'd be nice to mention the locking issues with reaching into a `NodeConnCB` from the main thread.  We could mention that `wait_until()` grabs the global lock before checking the predicate, or perhaps better still would be to add an example of acquiring the lock in this example before inspecting an object that is accessed from the other thread.